src/python/pants/backend/python/target_types.py

```diff
@@ -86,6 +86,60 @@ def value_or_global_default(self, python_setup: PythonSetup) -> Tuple[str, ...]:
         return python_setup.compatibility_or_constraints(self.value)
 
 
+class UnrecognizedResolveNamesError(Exception):
+    def __init__(
+        self,
+        unrecognized_resolve_names: list[str],
+        all_valid_names: Iterable[str],
+        *,
+        description_of_origin: str,
+    ) -> None:
+        # TODO(#12314): maybe implement "Did you mean?"
+        if len(unrecognized_resolve_names) == 1:
+            unrecognized_str = unrecognized_resolve_names[0]
+            name_description = "name"
+        else:
+            unrecognized_str = str(sorted(unrecognized_resolve_names))
+            name_description = "names"
+        super().__init__(
+            f"Unrecognized resolve {name_description} from {description_of_origin}: "
+            f"{unrecognized_str}\n\nAll valid resolve names: {sorted(all_valid_names)}"
```

Comment:
NB: the all_valid_names will be different when you're using the `resolve` field vs `generate-lockfiles --resolve`. In the latter case, it includes tool lockfiles. This is accurate: it's an error to use a tool resolve for user code.

But it could be confusing. I wonder if this error message should better explain this nuance? Something like:

> All valid resolve names (from `[python-setup].experimental_resolves_to_lockfiles`):

vs

> All valid resolve names (from `[python-setup].experimental_resolves_to_lockfiles` and all activated Python tools):

Reply:
Are tool resolves and user resolves conceptually in the same namespace?
Is the `black` tool resolve always named `black` so that all you have to do to replace the built-in lockfile is add an entry for `black` in `[python-setup].experimental_resolves_to_lockfiles`? Or would I add a user resolve (of any random name like `my_black_lockfile`) and set `[black].resolve = my_black_lockfile`?

In other words, how ambiguous might this be when a tool resolve is included in this list?

Reply:
A tool resolve is always its option scope, so `black`, `pytest`, `mypy_protobuf`. The lockfile path is set via `[black].lockfile` and so on, rather than `[python-setup].experimental_resolves_to_lockfiles` which is only for your own code. 

When you run `./pants generate-lockfiles`, we eagerly validate that your own resolve names do not conflict with any tool resolves.

So the only time the two types of resolves should really interact is when you run `./pants generate-lockfiles --resolve=black --resolve=my-custom-name`, which needs to be unambiguous. It also is currently an error if you set `pex_binary(resolve="black")` as that resolve needs to not be from a tool.

Does that make sense?